client/consensus/src/tests.rs

```diff
@@ -0,0 +1,775 @@
+    pub async fn start_orchestrator_aura_consensus_candidate_producer<B, C, SO, CIDP>(
+        slot_duration: SlotDuration,
+        client: C,
+        mut parachain_block_producer: Box<dyn ParachainConsensus<B>>,
+        sync_oracle: SO,
+        create_inherent_data_providers: CIDP,
+    ) where
+        B: BlockT,
+        C: SelectChain<B>,
+        SO: SyncOracle + Send,
+        CIDP: CreateInherentDataProviders<B, ()> + Send + 'static,
+        CIDP::InherentDataProviders: InherentDataProviderExt + Send,
+    {
+        let mut slots = Slots::new(
+            slot_duration.as_duration(),
+            create_inherent_data_providers,
+            client,
+        );
+
+        loop {
+            let slot_info = slots.next_slot().await;
+
+            if sync_oracle.is_major_syncing() {
+                continue;
+            }
+
+            let _ = parachain_block_producer
+                .produce_candidate(
+                    &slot_info.chain_head,
+                    Default::default(),
+                    &Default::default(),
+                )
+                .await;
+        }
+    }
+
+    // After all this boiler plate, tests start
+    #[tokio::test]
+    async fn authoring_blocks_but_producing_candidates_instead_of_calling_on_slot() {
+        let net = AuraTestNet::new(3);
+
+        let peers = &[
+            (0, Keyring::Alice),
+            (1, Keyring::Bob),
+            (2, Keyring::Charlie),
+        ];
+
+        let net = Arc::new(Mutex::new(net));
+        let mut import_notifications = Vec::new();
+        let mut aura_futures = Vec::new();
+
+        let mut keystore_paths = Vec::new();
+
+        // For each peer, we start an instance of the orchestratorAuraConsensus
+        // Only one of those peers will be able to author in each slot
+        // The tests finishes when we see a block higher than block 5 that has
+        // not being authored by us
+        for (peer_id, key) in peers {
+            let mut net = net.lock();
+            let peer = net.peer(*peer_id);
+            let client = peer.client().as_client();
+            let select_chain = peer.select_chain().expect("full client has a select chain");
+            let keystore_path = tempfile::tempdir().expect("Creates keystore path");
+            let keystore = Arc::new(
+                LocalKeystore::open(keystore_path.path(), None).expect("Creates keystore."),
+            );
+
+            keystore
+                .sr25519_generate_new(NIMBUS_KEY_ID, Some(&key.to_seed()))
+                .expect("Creates authority key");
+            keystore_paths.push(keystore_path);
+
+            let environ = DummyFactory(client.clone());
+            import_notifications.push(
+                client
+                    .import_notification_stream()
+                    .take_while(|n| {
+                        future::ready(!(n.origin != BlockOrigin::Own && n.header.number() < &5))
```

Comment:
This line is hard to understand, and it doesn't match the comment above. The comment says that 

```
// The tests finishes when we see a block higher than block 5 that has
// not being authored by us
```

But the condition is "a block lower than block 5 that has not being authored by us".

Removing one layer of negation can make it easier to understand, assuming I did it correctly:

```suggestion
                        future::ready(n.origin == BlockOrigin::Own || n.header.number() >= &5)
```

But I still need to understand this test.

Reply:
that is right, thank you

Reply:
actually the test is correct, is my comment which is wrong.

You have 3 authors, so you would expect to have a different author after 2 slots. I guess the magic number 5 has something to do with potential slot missing